suitcase/tiff/__init__.py

```diff
@@ -1,25 +1,312 @@
+        self._meta['metadata']['descriptors'] = defaultdict(dict)
+        self._stream_names = {}  # maps stream_names to each descriptor uids
+        self._files = {}  # map descriptor uid to file handle of tiff file
+        self._filenames = {}  # map descriptor uid to file names of tiff files
+        self._file_prefix = file_prefix
+        self._templated_file_prefix = ''
+        self._kwargs = kwargs
+
+
+    def start(self, doc):
+        '''Add `start` document information to the metadata dictionary.
+
+        This method adds the start document information to the metadata
+        dictionary. In addition it checks that only one `start` document is
+        seen.
+
+        Parameters:
+        -----------
+        doc : dict
+            The document dictionary associated with the start message.
+        '''
+
+        # raise an error if this is the second `start` document seen.
+        if 'start' in self._meta['metadata']:
+            raise RuntimeError(
+                "The serializer in suitcase.tiff expects documents from one "
+                "run only. Two `start` documents where sent to it")
+
+        # add the start doc to self._meta and format self._file_prefix
+        self._meta['metadata']['start'] = doc
+        self._templated_file_prefix = self._file_prefix.format(**doc)
+
+        # return the start document
+        return doc
+
+
+    def stop(self, doc):
+        '''Add `stop` document information to the metadata dictionary.
+
+        This method adds the stop document information to the metadata
+        dictionary. In addition it also creates the metadata '.json' file and
+        exports the metadata dictionary to it.
+
+        Parameters:
+        -----------
+        doc : dict
+            The document dictionary associated with the stop message.
+        '''
+        # add the stop doc to self._meta.
+        self._meta['metadata']['stop'] = doc
+
+        # open a json file for the metadata and add self._meta to it.
+        f = self.manager.open('run_metadata',
+                              f'{self._templated_file_prefix}meta.json', 'xt')
+        json.dump(self._meta, f)
+        self.artifacts['run_metadata'].append(f)
+        self._files['meta'] = f
+
+        # return the stop document
+        return doc
+
+
+    def descriptor(self, doc):
+        '''Add `descriptor` document information to the metadata dictionary.
+
+        This method adds the descriptor document information to the metadata
+        dictionary. In addition it also creates the file for data with the
+        stream_name given by the descriptor doc for later use.
+
+        Parameters:
+        -----------
+        doc : dict
+            The document dictionary associated with the descriptor message.
+
+        '''
+        # extract some useful info from the doc
+        stream_name = doc.get('name')
+        filename = self._templated_file_prefix + f'{stream_name}.tiff'
```

Comment:
Minor thing, but it's a bit weird to use `+` to concatenate the first two and `f''` to concatenate the second two. Might as well use `f''` for all of it.

Reply:
done